Feature/chung step3

kochungcheon opened the pull request:
# 구현 목표
1. 파일 / 폴더 이동
    - [x] 하위 모든 파일 / 폴더 이동
    - [x] 테스트 코드
 
2. 폴더 삭제
    - [x] 하위 모든 파일 / 폴더 삭제 (Hard Delete)
    - [x] 테스트 코드
    - [x] 삭제 된 파일 -> 로그 기록

## [삭제 로직]
- 폴더 삭제 로직의 경우 DFS 를 통해 구현 했습니다
- DFS 는 스택과 반복문을 통해 구현했습니다. 재귀를 통해 구현하게 되면 폴더 깊이가 매우 깊어졌을 때 문제가 생길 수 있기 때문입니다. 
- 메타 데이터만 삭제 되도록 정책을 수정하였습니다 
-  단, 실제 파일은 삭제 하지 않고 메타 데이터만 삭제 되게 됩니다
- 이를 통해, 스토리지와 데이터 베이스 사이 동기화 문제를 해결하였습니다. (메타데이터는 있는 데, 실제 파일은 삭제 되는 것과 같은 문제)

### [삭제 로그 기록]
- 파일 메타 데이터만 삭제하고 있기 때문에, 실제 파일의 삭제가 필요한 경우를 위해 만들었습니다. (저장소 메모리 관리 측면)
- 또한, 유저로부터 파일 삭제 로그 관련 문의가 왔을 때를 위해 만들었습니다. 

## [이동 로직]
- 파일/폴더는 논리적으로 연결되어 있습니다. 하위 파일/폴더는 부모 폴더의 Id를 가짐으로 연결 리스트와 유사하게 (실제는 트리겠지만) 관리 됩니다 그렇기에 parentId 만 바꾸는 방식을 통해 폴더 이동을 구현 했습니다

## [설계]

스토리지와 데이터 베이스로 나뉩니다

데이터 베이스에는 폴더와 파일의 메타 데이터가 저장됩니다

스토리지에는 실제 파일이 저장됩니다. 

폴더의 경우는 스토리지에 생성되지 않습니다. 논리적 관계만 데이터베이스에 존재합니다

이를 통해 저장 공간 절약, 유연한 구조라는 장점을 지닙니다 (예를 들어 파일 이동의 경우 상위 폴더 Id만 바꾸면 됩니다)

스토리지 구조도
basePath/
  ├── files1/
  ├── folders1/
  ├── files2/
  ├── folders2/
  └── ...

# 추후 구현할 사항

- 현재는 물리 저장소에 저장하고 있습니다. 서버 배포 시 해당 로직으로는 정상적으로 파일 API 가 동작하지 않음으로, S3 와 같은 저장소로 물리 저장소를 추후 대체하겠습니다.

kochungcheon committed:
feat: 파일 이동
- 폴더Id 변환으로 구현
트리 형태로 파일과 폴더를 관리함으로 폴더Id만 교환
모든 경로를 파일이 가지는 것 보다 효율적으로 관리 될 것이라 판단

- 적절한 메서드 추상화를 해야한다 느꼈습니다
Handler 메서드들이 다른 메서드들을 호출하는 형태로 구현하였습니다
추상화를 통해 특정 Service 에만 특화된 Handler 메서드를 만들게 되니, 유사한 기능이 필요한 Service가 존재 할때 메서드를 재활용 하지 못하는 문제가 생겼습니다

해당 문제가 발생한 메서드가 서로 다른 메서드들을 호출 하지 않게 바꾸었습니다

- 사용하지 않는 메서드 삭제

- 테스트 코드 작성
변경 사항 발생해도 테스트 코드 덕분에 안정적으로 프로그래밍 할 수 있었습니다

kochungcheon committed:
feat: 폴더 이동
- 폴더Id 변환으로 구현
트리 형태로 폴더를 관리함으로 폴더Id만 교환
모든 경로를 폴더가 가지는 것 보다 효율적으로 관리 될 것이라 판단

- 테스트 코드 작성

kochungcheon committed:
feat: feat: 폴더 삭제
- DFS 로직을 활용한 폴더 삭제
- Soft Delete를 하기로 하였음
파일/폴더 복구 문제와 서버 부하 고려

kochungcheon committed:
refactor: 폴더 조회 개선

kochungcheon committed:
feat: member 관련 기능
- user별 basePath를 만들어 주기 위해서

kochungcheon committed:
test: member 관련 test 생성

kochungcheon committed:
refactor : 사용자 로직 제거
- 폴더 관리를 위해 사용자 로직이 존재하는 당위를 찾지 못해 제거 하였습니다

kochungcheon committed:
refactor: 폴더 경로 로직 수정
- 스토리지 에는 실제 파일이 저장됩니다.

폴더의 경우는 스토리지에 생성되지 않습니다. 논리적 관계만 데이터베이스에 존재합니다
단일한 경로 아래서 폴더 및 파일 이 관리되도록 수정정하였습니다

kochungcheon committed:
chore: 코드 스멜 수정

kochungcheon committed:
chore: 코드 스멜 수정
- import 위치 변경

VSFe commented on src/main/java/com/c4cometrue/mystorage/file/FileService.java:
```diff
@@ -68,4 +72,28 @@ public CursorFileResponse getFiles(Long parentId, Long cursorId, Long userId, Pa
+	}
+
+	// 소프트 딜리트
+	public void deleteFile(FileMetadata fileMetadata) {
```
hard delete는 시행하지 않나요?

kochungcheon replied:
> hard delete는 시행하지 않나요?

~우선 소프트 딜리트를 진행하고 이후 일정 시간 마다 소프트 딜리트된 파일들을 하드 딜리트하는 것으로 바꾸었습니다!~

kochungcheon replied:
> hard delete는 시행하지 않나요?

hard delete 수행으로 바꾸었습니다!

VSFe commented on src/main/java/com/c4cometrue/mystorage/storage/StorageFacadeService.java:
```diff
@@ -36,15 +38,36 @@ private CursorMetaResponse handleFolderFirstStrategy(Long parentId, Long cursorI
+	// softDelete
+	public void deleteFolderContents(FolderMetadata folderMetadata) {
+		// 하위 폴더 재귀적으로 약한 삭제
+		folderService.findAllBy(folderMetadata.getId())
```
중간에 폴더 삭제가 실패하면, 어떻게 처리해야 하나요?

kochungcheon replied:
> 중간에 폴더 삭제가 실패하면, 어떻게 처리해야 하나요?

최상단 메서드에 트랜잭션을 달아서 원자적으로 처리되게 하였습니다

최상단 메서드인 deleteFolderContents(long folderId, long userId)와 하위 메서드인 deleteFolderContents(FolderMetadata folderMetadata) 중 어디에 트랜잭션을 달아야 할 지 고민하였습니다

하위 메서드에 달게 되면 다른 트랜잭션은 실패해도 부분 삭제가 가능하지만, 사용자 입장에서는 어떤 폴더/파일이 부분 삭제 됬는 지 파악하기 힘들기 때문에 사용자 경험을 해칠 우려가 있다는 판단이였습니다.

VSFe commented on src/main/java/com/c4cometrue/mystorage/storage/StorageFacadeService.java:
```diff
@@ -36,15 +38,36 @@ private CursorMetaResponse handleFolderFirstStrategy(Long parentId, Long cursorI
+	public void deleteFolderContents(FolderMetadata folderMetadata) {
+		// 하위 폴더 재귀적으로 약한 삭제
+		folderService.findAllBy(folderMetadata.getId())
+			.forEach(this::deleteFolderContents);
```
DFS를 사용했을 때, 메모리 부하 관점에선 문제가 없나요? (지적이 아니라 개인적인 생각을 작성해 주세요.)

kochungcheon replied:
> DFS를 사용했을 때, 메모리 부하 관점에선 문제가 없나요? (지적이 아니라 개인적인 생각을 작성해 주세요.)

폴더 삭제 로직 BFS 와 DFS 를 고려했습니다
DFS 는 BFS 와 달리 level 별 모든 데이터를 가지고 있지 않아도 된다는 장점을 가져 메모리 부하 관점에서 이점을 가진다 생각합니다

VSFe replied:
특정 유저가 회원탈퇴를 하기 위해 루트 폴더에서 폴더 삭제를 구현했다고 가정해 봅시다.
결국 DFS이기 때문에 Folder의 depth가 깊어지면 깊어질수록 Recursive Function의 depth도 커지게 되겠죠.
(막말로 폴더가 한 100중으로 되어 있고, 각각에 대해서 파일도 상당히 많다고 가정해 봅시다. 어떻게 될까요?)

Recursive Function을 최적화 하기 위해선 Tail-Recursion 구조가 되어야 하는데, 여기에선 그렇지 않으니 최적화가 불가능하겠죠?

kochungcheon replied:
> 특정 유저가 회원탈퇴를 하기 위해 루트 폴더에서 폴더 삭제를 구현했다고 가정해 봅시다. 결국 DFS이기 때문에 Folder의 depth가 깊어지면 깊어질수록 Recursive Function의 depth도 커지게 되겠죠. (막말로 폴더가 한 100중으로 되어 있고, 각각에 대해서 파일도 상당히 많다고 가정해 봅시다. 어떻게 될까요?)
> 
> Recursive Function을 최적화 하기 위해선 Tail-Recursion 구조가 되어야 하는데, 여기에선 그렇지 않으니 최적화가 불가능하겠죠?

앗 덕분에 Tail-Recursion에 대한 공부를 하였습니다. 감사합니다. 

스택과 반복문을 이용한 DFS 로직으로 해당 문제를 해결하였습니다.

VSFe commented on src/main/resources/application.yml:
```diff
@@ -6,7 +6,7 @@ spring:
   jpa:
     hibernate:
-      ddl-auto: update
+      ddl-auto: create-drop
```
?

kochungcheon replied:
> ?

앗.. update로 다시 변경하였습니다. 테이블 구조 변경하다 생긴 미스..

VSFe reviewed (commented):
생각보다 수정할 사항이 많진 않네요~

kochungcheon committed:
refactor : yml create-drop -> update

kochungcheon committed:
refactor : @Tx 추가
- 최상단 메서드인 deleteFolderContents(long folderId, long userId)에 추가
- 삭제시 원자적으로 처리하는 것이 사용자 경험에 더 좋다는 판단

kochungcheon committed:
feat: metadataBaseEntity 추가
- 데이터 변경시(삭제, 업데이트) 추적하기 위해서

kochungcheon committed:
feat : hard delete
- 폴더 삭제를 할 시 삭제 된 파일 / 폴더는 deletemeradata 에 저장되게 수정 하였습니다
폴더 삭제 시 가장 신경 썼던 부분은 롤백 상황이였습니다. 물리적 파일이 삭제 되었는 데
데이터 베이스에는 롤백으로 인해 정보가 남아있을 수 있기 때문입니다.
이를 해결하고자 삭제 시 바로 삭제 되는 게 아니라 deletemeradata 를 거쳐 삭제 되게 합니다
deletemeradata 에는 일반 유저가 접근 할 수 없기 때문에 설령 롤백이 될 지라도 상관없습니다. 삭제 되지 않은 메타데이터는 다음날 삭제 시도하면 되기 때문입니다
- 매일 오전 3시 deletemeradata 에서는 hard delete 가 수행됩니다
- deletemeradata 을 삭제 하는 로직에 커서 기반 페이징 적용
deletemeradata 테이블은 필연적으로 데이터가 많이 쌓이는 구조입니다. 그렇기에 페이징을 통해서 100개씩 끊어서 처리하고자 하였습니다.

kochungcheon committed:
test: add deleteMetadataTest
- DeleteMetadata 관련 테스트
- File 관련 테스트 추가
- Folder 관련 테스트 추가

kochungcheon committed:
test: fileDataHandler 테스트 수정

kochungcheon committed:
chore: code smell 수정

VSFe commented on src/main/java/com/c4cometrue/mystorage/util/PagingUtil.java:
```diff
@@ -4,10 +4,12 @@
 import org.springframework.data.domain.Pageable;
 
 public class PagingUtil {
```
Spring Data의 PageRequest, Pageable은 기본적으로 Offset-Based Paging을 하고 있습니다.
지금 작성한 코드는 억지로 Cursor-Based Paging을 하기 위해 구성한 것 처럼 보이는데, 사실 그냥 Pageable을 안 쓰는게 낫지 않을까 생각이 드네요.

VSFe commented on src/main/java/com/c4cometrue/mystorage/deletedmetadata/DeleteMetadataScheduledTasks.java:
```diff
@@ -0,0 +1,47 @@
+			// 삭제된 파일들은 데이터 베이스에서도 지워준다
+			deletedMetadataRepository.deleteAllById(metadataIds);
+
+			cursorId = deletedMetadataList.get(deletedMetadataList.size() - 1).getId();
```
(고치라는 건 아니고, 한 번 생각해 보세요.)
K개씩 끊어서 페이징을 한다고 가정해보면, 현재 로직으로는 불필요한 1회의 추가 쿼리 요청을 더 할 가능성이 높아요.
(ex. K = 200인 상황에서, 13개만 검색된 경우)
어떻게 작성하는게 가장 효율적일까요?

kochungcheon replied:
> (고치라는 건 아니고, 한 번 생각해 보세요.) K개씩 끊어서 페이징을 한다고 가정해보면, 현재 로직으로는 불필요한 1회의 추가 쿼리 요청을 더 할 가능성이 높아요. (ex. K = 200인 상황에서, 13개만 검색된 경우) 어떻게 작성하는게 가장 효율적일까요?

아 K=200 일 시 이거 보다 작으면 추가 쿼리 요청을 안하게 수정하면 되는 군요!!!! 감사합니다!

VSFe commented on src/main/java/com/c4cometrue/mystorage/deletedmetadata/DeleteMetadataScheduledTasks.java:
```diff
@@ -0,0 +1,47 @@
+	private final DeletedMetadataRepository deletedMetadataRepository;
+
+	@Transactional
+	@Scheduled(cron = "0 0 3 * * ?")
```
매일 1회면 상당히 데이터가 많이 쌓이겠네요.

kochungcheon replied:
> 매일 1회면 상당히 데이터가 많이 쌓이겠네요.

앗 그렇습니다. 전체가 하나의 컴포넌트로 묶여있다보니 이런 문제가 생기네요
서버에 부담이 많이 가는 로직이라 판단하여 사용자 들이 많이 안 사용하는 시간에 스케줄을 돌렸습니다.

확실히 컴포넌트를 분리해야 할 이유가 느는 군요

VSFe commented on src/main/java/com/c4cometrue/mystorage/deletedmetadata/DeleteMetadataScheduledTasks.java:
```diff
@@ -0,0 +1,47 @@
+
+@Service
+@RequiredArgsConstructor
+public class DeleteMetadataScheduledTasks {
```
Scheduled Task가 같이 있는건 썩 좋아보이진 않네요. (물론 지금 구조상 어쩔 수 없기에 패스하지만, Scheduled Job만 다루는 컴포넌트를 분리할 수 있다면 훨씬 나았을 것 같아요.)

kochungcheon replied:
> Scheduled Task가 같이 있는건 썩 좋아보이진 않네요. (물론 지금 구조상 어쩔 수 없기에 패스하지만, Scheduled Job만 다루는 컴포넌트를 분리할 수 있다면 훨씬 나았을 것 같아요.)

앗 그렇습니다 컴포넌트 분리가 확실히 좋아보이네요!

VSFe commented on src/main/java/com/c4cometrue/mystorage/deletedmetadata/DeleteMetadataScheduledTasks.java:
```diff
@@ -0,0 +1,47 @@
+
+		List<DeletedMetadata> deletedMetadataList = deletedMetadataRepository.findAllWithCursor(cursorId, pageable);
+
+		while (!deletedMetadataList.isEmpty()) {
```
현재 로직은 Hard Delete를 수행하다 1번이라도 예외가 발생하면 즉시 중단됩니다.
삭제해야 할 파일이 20,000개인데 10개 삭제했을 때 예외가 발생하면 19,990개의 파일은 그대로 남아있겠죠.

거기에, 예외가 발생한 파일은 결국 삭제되지 않으므로 계속 예외를 내놓게 될 것이고, 결과적으로 서비스 장애의 원인이 될 것으로 보여요. 로직에 대한 검토가 필요해 보입니다.

kochungcheon replied:
> 현재 로직은 Hard Delete를 수행하다 1번이라도 예외가 발생하면 즉시 중단됩니다. 삭제해야 할 파일이 20,000개인데 10개 삭제했을 때 예외가 발생하면 19,990개의 파일은 그대로 남아있겠죠.
> 
> 거기에, 예외가 발생한 파일은 결국 삭제되지 않으므로 계속 예외를 내놓게 될 것이고, 결과적으로 서비스 장애의 원인이 될 것으로 보여요. 로직에 대한 검토가 필요해 보입니다.

한번 예외가 생긴 파일은 계속 예외가 생길 수 있군요. 아.. 재 검토 해보겠습니다! 감사합니다!

VSFe commented on src/main/java/com/c4cometrue/mystorage/deletedmetadata/DeletedMetadata.java:
```diff
@@ -0,0 +1,42 @@
+
+	@PreUpdate
+	public void prePersist() {
+		deletedDate = LocalDate.now();
```
날짜 정보만 저장하는 이유가 있을까요?

kochungcheon replied:
> 날짜 정보만 저장하는 이유가 있을까요?
로깅을 위해 저장하였습니다!
매일 1회 데이터들을 지우게 되는 데, 이슈가 발생 시 삭제 되지 않은 파일들을 확실히 구분할려고요!

VSFe reviewed (commented):
네넴띤

kochungcheon committed:
refactor: 스케줄러를 통한 삭제 제거

kochungcheon committed:
refactor : 파일/폴더 삭제

kochungcheon committed:
refactor: DFS 재귀 -> 스택을 활용한 로직

kochungcheon committed:
refactor: time 변경

kochungcheon committed:
feat: 파일 삭제 로그 기록
- 로그를 이용하면 실제 파일을 지워야 할 필요가 있을 때 처리가 가능하여 추가
- 사용자로부터 삭제 기록 요청 시, 확인을 위해 추가

kochungcheon committed:
test: 파일 삭제 로그 테스트 코드 추가

kochungcheon committed:
refactor: 삭제 날짜 추가 in 로그

kochungcheon committed:
chore: CSRF -> LF

kochungcheon committed:
chore: import order
